lahja/trio/endpoint.py

```diff
@@ -0,0 +1,913 @@
+import asyncio
+import collections
+import itertools
+import logging
+import pathlib
+import pickle
+from typing import (
+    Any,
+    AsyncGenerator,
+    AsyncIterator,
+    Awaitable,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+import weakref
+
+from async_generator import asynccontextmanager
+import trio
+import trio_typing
+
+from lahja import constants
+from lahja.base import (
+    BaseEndpoint,
+    BaseRemoteEndpoint,
+    ConnectionAPI,
+    EndpointAPI,
+    RemoteEndpointAPI,
+    TResponse,
+    TStreamEvent,
+    TSubscribeEvent,
+    TWaitForEvent,
+)
+from lahja.common import (
+    BaseEvent,
+    BaseRequestResponseEvent,
+    Broadcast,
+    BroadcastConfig,
+    ConnectionConfig,
+    Message,
+    Msg,
+    Subscription,
+    should_endpoint_receive_item,
+)
+from lahja.exceptions import (
+    ConnectionAttemptRejected,
+    RemoteDisconnected,
+    UnexpectedResponse,
+)
+from lahja.typing import ConditionAPI, RequestID
+
+
+class TrioConnection(ConnectionAPI):
+    logger = logging.getLogger("lahja.trio.Connection")
+
+    def __init__(self, socket: trio.SocketStream) -> None:
+        self._socket = socket
+        self._msg_send_channel, self._msg_receive_channel = cast(
+            Tuple[trio.abc.SendChannel[Message], trio.abc.ReceiveChannel[Message]],
+            trio.open_memory_channel(100),
+        )
+        self._lock = trio.Lock()
+        self._running = trio.Event()
+        self._stopped = trio.Event()
+        super().__init__()
+
+    def __str__(self) -> str:
+        return f"TrioConnection[{self._socket}]"
+
+    def __repr__(self) -> str:
+        return f"<{self}>"
+
+    #
+    # Lifecycle API
+    #
+    @property
+    def is_running(self) -> bool:
+        return not self.is_stopped and self._running.is_set()
+
+    @property
+    def is_stopped(self) -> bool:
+        return self._stopped.is_set()
+
+    async def wait_started(self) -> None:
+        await self._running.wait()
+
+    async def wait_stopped(self) -> None:
+        await self._stopped.wait()
```

Comment:
I spent some time trying to make a `Runnable` abstraction but it ended up making things feel really disconnected complex to understand.  Maybe worth trying for again once this code has settled.